servant-server/src/Servant/Server/Internal.hs

```diff
@@ -134,8 +135,8 @@ processMethodRouter :: forall a. ConvertibleStrings a B.ByteString
                     -> Maybe [(HeaderName, B.ByteString)]
                     -> Request -> RouteResult Response
 processMethodRouter handleA status method headers request = case handleA of
-  Nothing -> failWith UnsupportedMediaType
-  Just (contentT, body) -> succeedWith $ responseLBS status hdrs bdy
+  Nothing -> FailFatal err406
```

Comment:
Also supported would be 
```
     Put '[Foo] A
:<|> Put '[Bar] B
```
Which as I've mentioned elsewhere I think is a bad idea, since content-type should not be allowed to affect the nature (type) of the resource, only it's presentation.